src/main/java/org/HdrHistogram/StripedWriterReaderPhaser.java

```diff
@@ -0,0 +1,227 @@
+ */
+
+package org.HdrHistogram;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLongArray;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * {@link StripedWriterReaderPhaser} instances provide an asymmetric means for synchronizing the execution of
+ * wait-free "writer" critical sections against a "reader phase flip" that needs to make sure no writer critical
+ * sections that were active at the beginning of the flip are still active after the flip is done. Multiple writers
+ * and multiple readers are supported.
+ * <p>
+ * While a {@link StripedWriterReaderPhaser} can be useful in multiple scenarios, a specific and common use case is
+ * that of safely managing "double buffered" data stream access in which writers can proceed without being
+ * blocked, while readers gain access to stable and unchanging buffer samples
+ * <blockquote>
+ * NOTE: {@link StripedWriterReaderPhaser} writers are wait-free on architectures that support wait-free atomic
+ * increment operations. They remain lock-free (but not wait-free) on architectures that do not support
+ * wait-free atomic increment operations.
+ * </blockquote>
+ * {@link StripedWriterReaderPhaser} "writers" are wait free, "readers" block for other "readers", and
+ * "readers" are only blocked by "writers" whose critical section was entered before the reader's
+ * {@link StripedWriterReaderPhaser#flipPhase()} attempt.
+ * <p>
+ * When used to protect an actively recording data structure, the assumptions on how readers and writers act are:
+ * <ol>
+ * <li>There are two sets of data structures ("active" and "inactive")</li>
+ * <li>Writing is done to the perceived active version (as perceived by the writer), and only
+ *     within critical sections delineated by {@link StripedWriterReaderPhaser#writerCriticalSectionEnter}
+ *     and {@link StripedWriterReaderPhaser#writerCriticalSectionExit}).</li>
+ * <li>Only readers switch the perceived roles of the active and inactive data structures.
+ *     They do so only while under readerLock(), and only before calling flipPhase().</li>
+ * </ol>
+ * When the above assumptions are met, {@link StripedWriterReaderPhaser} guarantees that the inactive data structures are not
+ * being modified by any writers while being read while under readerLock() protection after a flipPhase()
+ * operation.
+ *
+ *
+ *
+ */
+public class StripedWriterReaderPhaser {
+    private final int stripes;
+
+    // TODO: prevent false sharing
+    private final AtomicLongArray startEpoch;
+    private final AtomicLongArray evenEndEpoch;
+    private final AtomicLongArray oddEndEpoch;
+
+    private final ReentrantLock readerLock = new ReentrantLock();
+    private final long stride;
+
+    public StripedWriterReaderPhaser(final int stripes) {
+        // stripes must be a power of 2.
+        assert stripes > 0 && (stripes & (stripes - 1)) == 0;
+        this.stripes = stripes;
+        startEpoch = new AtomicLongArray(stripes);
```

Comment:
I suspect that to really benefit from the avoided collisions you should need to wrap each long value into 2 cache lines (prefetcher size) to avoid false sharing between them.

Reply:
Since the values are in an array of atomic longs, making the array 16 times as large and adjusting the math to only using every 16th element will do the trick (when targetting 128 byte spacing, which is what x86 would need to avoid some of the prefetching behaviors).